client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+        private CompletableFuture<ByteBuffer> outstandingPromise = CompletableFuture.completedFuture(null);
+        private ByteBuffer resultBuffer;
+        private long offset;
+
+        /**
+         * Gets the next offset for reading purposes.
+         */
+        public long getOffset() {
+            return offset;
+        }
+
+        /**
+         * Resets the read state in preparation for a new read.
+         * @param offset the offset to prepare for.
+         */
+        public CompletableFuture<ByteBuffer> reset(long offset) {
+            if (!this.outstandingPromise.isDone()) {
+                this.outstandingPromise.cancel(false);
+            }
+            this.outstandingPromise = new CompletableFuture<>();
+            this.headerReadingBuffer.clear();
+            this.resultBuffer = null;
+            this.offset = offset;
+            return this.outstandingPromise;
+        }
+
+        /**
+         * Updates the read state based on the given {@link SegmentRead}.
+         * Note that the {@code segmentRead} is progressively consumed by successive event reads.
+         *
+         * @param segmentRead a read response.
+         */
+        public void update(SegmentRead segmentRead) {
+            checkState(!isSuccessful());
+            verifyIsAtCorrectOffset(segmentRead);
+
+            if (resultBuffer == null) {
+                // consume header data from the SegmentRead data buffer
+                if (headerReadingBuffer.hasRemaining()) {
+                    assert headerReadingBuffer.limit() == WireCommands.TYPE_PLUS_LENGTH_SIZE;
+                    offset += fill(headerReadingBuffer, segmentRead.getData());
+                }
+
+                if (!headerReadingBuffer.hasRemaining()) {
+                    // header is ready
+                    headerReadingBuffer.flip();
+                    int type = headerReadingBuffer.getInt();
+                    int length = headerReadingBuffer.getInt();
+                    if (type != WireCommandType.EVENT.getCode()) {
+                        throw new InvalidMessageException("Event was of wrong type: " + type);
+                    }
+                    if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                        throw new InvalidMessageException("Event of invalid length: " + length);
+                    }
+                    resultBuffer = ByteBuffer.allocate(length);
+                }
+            }
+
+            if (resultBuffer != null) {
+                // consume result data from the SegmentRead data buffer
+                if (resultBuffer.hasRemaining()) {
+                    offset += fill(resultBuffer, segmentRead.getData());
+                }
+
+                if (!resultBuffer.hasRemaining()) {
+                    // result is ready
+                    resultBuffer.flip();
+                    complete();
```

Comment:
This is invoking a callback under lock. 
This violates lock ordering. All locks should proceed from further from the connection down to the connection object. If there is a thread going the other way it cannot pass through these locks. This is why implementing this change correctly is non trivial.

Reply:
The lock in question is `$lock` on `AsyncSegmentEventReaderImpl`.  Yes, the lock could be acquired either by the caller (to control the reader) or by the connection (to deliver results), both of which mutate the internal state.   Can you think of a practical problem with this?

Reply:
You should never complete a Future while holding a lock. You never know what downstream callbacks will be invoked, in which order or whether they will be invoked synchronously or asynchronously. This can easily lead to deadlocks.

Reply:
I understand these general concerns but you're suggesting that the code be significantly more complex to solve a problem that doesn't exist w.r.t. the actual users of this class, i.e. `EventStreamReaderImpl`.

Reply:
EventStreamReaderImpl can and will change in the future. It cannot be assumed that it will have intimate knowledge of this class. Every class should be correct on it's own without having to have knowledge of the rest of the system. Otherwise it becomes very difficult to, perform code reviews, test components in isolation, or have confidence the system is correct.

Reply:
> The lock in question is `$lock` on `AsyncSegmentEventReaderImpl`.  Yes, the lock could be acquired either by the caller (to control the reader) or by the connection (to deliver results), both of which mutate the internal state.   Can you think of a practical problem with this?

As an example of why that may be a problem, imagine the netty thread holds a lock. This would deadlock our reads if we invoked some other netty function that depended on the same lock.

Similarly the same problem occurs inside of EventStreamReaderImpl. You have the member readCompletionQueue which is annotated that it is guarded by a lock. This will generate a compile error if you enable the static analysis in initliJ. But if someone were to see that and fix that bug by guarding it by the lock, the threads would deadlock.

Reply:
Good find, the `readCompletionQueue` should not be guarded in that way.

Reply:
Please follow our standard thread safety rules. These are on the wiki, but a presentation that might be helpful is here: 
https://drive.google.com/open?id=1IhOwnXynKtgqXgMpj5SSdF0uHWCYYQ6fnD_ypAN25No
I put that together a while ago for the team.